crates/data-generator/src/blocks/output.rs

```diff
@@ -0,0 +1,628 @@
+            let props_code = gen_props_code(self.0);
+
+            tokens.extend(quote! {
+                pub mod #block_mod_name {
+                    #block_code
+                    #props_code
+                }
+            });
+        }
+    }
+
+    fn gen_block_code(block: &Block) -> TokenStream {
+        let mut tokens = TokenStream::new();
+        super::block_struct::BlockStruct::from(block).to_tokens(&mut tokens);
+        super::block_convert::BlockConvert::from(block).to_tokens(&mut tokens);
+        tokens
+    }
+
+    fn gen_props_code(block: &Block) -> TokenStream {
+        let mut tokens = TokenStream::new();
+        for prop in block.properties.iter() {
+            let def_prop = block
+                .default_state
+                .properties
+                .iter()
+                .find(|def_prop| prop.name_sc == def_prop.field)
+                .unwrap();
+            let prop = prop.clone();
+
+            super::prop_enum::PropertyEnum::from(&prop).to_tokens(&mut tokens);
+            super::prop_default::PropertyDefault::from(def_prop).to_tokens(&mut tokens);
+            super::prop_convert::PropertyConvert::from(prop).to_tokens(&mut tokens);
+        }
+        tokens
+    }
+}
+
+mod block_struct {
+    use proc_macro2::{Delimiter, Group, Ident, Punct, Spacing, TokenStream};
+    use quote::{quote, ToTokens, TokenStreamExt as _};
+
+    use super::StringExt as _;
+    use crate::blocks::model::Block;
+
+    pub struct BlockStruct {
+        pub properties: Vec<BlockStructField>,
+    }
+
+    impl From<&Block> for BlockStruct {
+        fn from(block: &Block) -> Self {
+            let properties = block
+                .properties
+                .iter()
+                .map(|block_prop| BlockStructField {
+                    prop_name: block_prop.name_sc.to_ident(),
+                    prop_enum_name: block_prop.name_pc.to_ident(),
+                })
+                .collect();
+
+            Self { properties }
+        }
+    }
+
+    impl ToTokens for BlockStruct {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            tokens.extend(quote! {
+                #[derive(Debug, Default, Eq, PartialEq, Copy, Clone, Hash)]
+                pub struct Block
+            });
+
+            if self.properties.is_empty() {
+                // "pub struct LavaCauldron;"
+                tokens.append(Punct::new(';', Spacing::Alone));
+            } else {
+                let mut stream = TokenStream::new();
+                self.properties
+                    .iter()
+                    .for_each(|prop| prop.to_tokens(&mut stream));
+
+                // "pub struct Lectern { pub has_book: HasBook, }"
+                tokens.append(Group::new(Delimiter::Brace, stream));
+            }
+        }
+    }
+
+    pub struct BlockStructField {
+        pub prop_name: Ident,      // honey_level
+        pub prop_enum_name: Ident, // HoneyLevel
+    }
```

Comment:
These comments all over the place only made it harder to comprehend. One of the solutions I see is to have names of distinct things as `struct`s that encapsulate the handling details as an implementation detail.

For example, here we could get by with a single `PropertyName`, which we'd use like this:
```rs
impl ToTokens for BlockStructField {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        // This naming is lame but we can figure something out later.
        let field_name = self.prop_name.field_name();
        let enum_name = self.prop_name.enum_name();

        // "pub open: Open,"
        tokens.extend(quote!(pub #field_name: #enum_name,))
    }
}
```

Reply:
On top of that, this particular struct is probably not worth being extracted from `BlockStruct`.

The guideline for me was that functionality taking more than 5-10 lines should be extracted into a separate struct, which (1) proved too short to be extracted, and (2) no longer the case after all the changes to the original code of `BlockStructField`.

Reply:
`crate::name::Name`